### PR TITLE
C++: Implement guards logic for switch statements

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR is the follow-up to https://github.com/github/codeql/pull/15941 that addresses this comment I made in that PR:
> It's still possible to do more in this area. For example, GuardCondition.comparesEq still doesn't reason about switch statements since the interface on that predicate can't really support it (its columns include `(Expr left, Expr right, int k)` which holds if `left == right + k`, but the case label in a switch isn't an expression). So such changes are probably best done in a follow-up PR.

Commit-by-commit review recommended.

This PR only does the required API additions to support `GuardCondition.comparesEq` for `switch` statements. We also have a predicate `GuardCondition.comparesLt` which needs a similar treatment. I'm delaying that until a future PR

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 1]
```diff
@@ -572,52 +613,96 @@
...
   /* (x is true => (left == right + k)) => (!x is false => (left == right + k)) */
-  exists(boolean isFalse | testIsTrue = isFalse.booleanNot() |
-    compares_eq(test.(LogicalNotInstruction).getUnary(), left, right, k, areEqual, isFalse)
+  exists(AbstractValue dual | value = dual.getDualValue() |
+    compares_eq(test.(LogicalNotInstruction).getUnary(), left, right, k, areEqual, dual)
+  )
+}
+
+/** Holds if `op == k` is `areEqual` given that `test` is `testIsTrue`. */
+private predicate compares_eq(
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:640]: Fixed in 98396d1e9a30d3ff3ffa42c97b6d0eee4b95ac4f

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll [hunk 2]
```diff
@@ -90,6 +90,15 @@
    * Gets the largest value of the switch expression for which control will flow along this edge.
    */
   final string getMaxValue() { result = maxValue }
+
+  /**
+   * Gets the unique value of ther switch expression for which control will
+   * flow along this edge, if any.
+   */
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll]: Fixed in a56b5cd4a512a8b9b81b7b39e971b551c70c9606

**Comment by jketema** [comment]
I see some alert changes in DCA. Are those correct?

**Comment by MathiasVP** [comment]
> I see some alert changes in DCA. Are those correct?

I _thought_ they were correct, but now that I look slightly closer I've spotted a problem. Thanks for making me double check them!

**Comment by MathiasVP** [comment]
I should've paid more attention to the test changes. They revealed the same problem >.<

**Review by jketema — APPROVED** [review]
LGTM if DCA is happy this time around

**Comment by MathiasVP** [comment]
DCA looks uneventful now (🎉 I guess)

**Pull request merged**